simulariumio/readers/readdy_trajectory_reader.py

```diff
@@ -0,0 +1,199 @@
+            n = 0
+            for i in range(int(agent_data["n_agents"][t])):
+                if traj.species_name(agent_data["type_ids"][t][i]) not in ignore_types:
+                    n += 1
+            n_filtered_particles_per_frame[t] = n
+        # filter particle data to remove ignored types
+        max_n_particles = int(np.amax(n_filtered_particles_per_frame))
+        result = {
+            "times": agent_data["times"],
+            "viz_types": agent_data["viz_types"],
+            "n_agents": n_filtered_particles_per_frame,
+            "unique_ids": -1 * np.ones((totalSteps, max_n_particles)),
+            "type_ids": np.zeros((totalSteps, max_n_particles)),
+            "positions": np.zeros((totalSteps, max_n_particles, 3)),
+            "radii": np.ones(shape=(totalSteps, max_n_particles)),
+        }
+        for t in range(len(agent_data["n_agents"])):
+            n = 0
+            for i in range(int(agent_data["n_agents"][t])):
+                type_name = traj.species_name(agent_data["type_ids"][t][i])
+                if type_name in ignore_types or n >= n_filtered_particles_per_frame[t]:
+                    continue
+                result["unique_ids"][t][n] = agent_data["unique_ids"][t][i]
+                result["type_ids"][t][n] = agent_data["type_ids"][t][i]
+                result["positions"][t][n] = agent_data["positions"][t][i]
+                result["radii"][t][n] = agent_data["radii"][t][i]
+                n += 1
+        return result
+
+    def _group_particle_types(
+        self, agent_data: Dict[str, Any], traj: Any, type_grouping: Dict[str, List[str]]
+    ) -> Tuple[Dict[str, Any], Dict[str, Dict[str, str]]]:
+        """
+        Group ReaDDy particle types by assigning them to new group type IDs
+        """
+        # map ReaDDy ID to new group ID
+        group_mapping = {}
+        type_map = traj.particle_types
+        type_mapping = {}
+        i = int(np.amax(agent_data["type_ids"])) + 1
+        for group_type in type_grouping:
+            type_mapping[str(i)] = {"name": group_type}
+            for readdy_type in type_grouping[group_type]:
+                if readdy_type in type_map:
+                    group_mapping[type_map[readdy_type]] = i
+                else:
+                    log.warning(
+                        f"type {readdy_type}, which was provided in the type_grouping, "
+                        "doesn't exist in the ReaDDy model"
+                    )
+            i += 1
+        # assign group ID to each particle of a type in the group
+        for t in range(len(agent_data["n_agents"])):
+            for n in range(int(agent_data["n_agents"][t])):
+                readdy_id = agent_data["type_ids"][t][n]
+                if readdy_id in group_mapping:
+                    agent_data["type_ids"][t][n] = group_mapping[readdy_id]
+        return (agent_data, type_mapping)
+
+    def read(self, data: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
It would go a long way to documenting this api if you could change these `Any`s (both the input and the return value) to some kind of more specific type definition.  I don't know how feasible that is, or if this is part of the api that users would need to worry about.

Reply:
there's going to be different types of data in this `data` object for different simulators, often fields won't be reused, for others they might be in a different format (see https://github.com/allen-cell-animated/simulariumio/blob/8f07c2c1fcd5a49c3f22e6f6b5dacb48be5359a6/simulariumio/converter.py#L49 for all the different simulators). I could try defining types for them, but I'm not sure how to handle inheritance with those type definitions. It seems like there isn't much overlap, so the definitions would be pretty complex and I'm not sure how useful they would be. Is there another way I should do this?

Reply:
When I see that docstring in converter.py showing all the different data for different simulators, it makes me think you need something like separate ReaddyConverter, PhysiCellConverter, CytosimConverter classes.  The generic Converter could discern the type it is dealing with and then forward the work to those classes.